<a href="https://colab.research.google.com/github/mkernik/Bbox_assignment/blob/main/RutgersPlaceNames.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
from google.colab import files

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
%cd /content/drive/My Drive/MetadataSprint/

/content/drive/My Drive/MetadataSprint


In [ ]:
df = pd.read_csv("RutgersSpatialCoverage.csv")

In [ ]:
place_names = dict(zip(df.geoportal_id, df.SpatialCoverage))
print(place_names)

{'rutgers-lib:30173': 'New Jersey', 'rutgers-lib:30182': 'New Jersey', 'rutgers-lib:30174': 'New Jersey', 'rutgers-lib:30183': 'New Jersey', 'rutgers-lib:30175': 'New Jersey', 'rutgers-lib:30184': 'New Jersey', 'rutgers-lib:30176': 'New Jersey', 'rutgers-lib:30185': 'New Jersey', 'rutgers-lib:30177': 'New Jersey', 'rutgers-lib:36695': 'Port Monmouth|Horseshoe Cove|Spermaceti CoveMonmouth County|Keyport, New Jersey|Monmouth County|Union Beach, New Jersey|Monmouth County|Keansburg, New Jersey|Monmouth County|Middletown Township, New JerseyMonmouth County|Atlantic Highlands, New Jersey|Monmouth County|Highlands, New Jersey|Middlesex County, New Jersye|Perth Amboy, New Jersey|Middlesex County, New Jersye|South Amboy, New Jersey|Monmouth County|Sandy Hook, New Jersey|New Jersey', 'rutgers-lib:36694': 'Ocean County|Little Egg Harbor, New Jersey|New Jersey', 'rutgers-lib:30186': 'New Jersey', 'rutgers-lib:30178': 'New Jersey', 'rutgers-lib:30187': 'New Jersey', 'rutgers-lib:30179': 'New Jerse

In [ ]:
df2 = pd.read_csv("NJCountyBBox_simple.csv")

In [ ]:
df2.head()
NJ_counties_dict = dict(zip(df2.CountyName, df2.BoundingBox))
print(NJ_counties_dict)

{'Atlantic': '-74.985271,39.288558,-74.232062,39.729775', 'Bergen': '-74.272226,40.76159,-73.893628,41.133714', 'Burlington': '-75.059913,39.541198,-74.389708,40.183651', 'Camden': '-75.139832,39.608352,-74.736216,39.996081', 'Cape May': '-75.168462,38.788657,-74.46493,39.324375', 'Cumberland': '-75.444751,39.056179,-74.851794,39.568746', 'Essex': '-74.377391,40.674368,-74.112787,40.90889', 'Gloucester': '-75.429061,39.514967,-74.87726,39.888317', 'Hudson': '-74.166087,40.64199,-73.984739,40.823582', 'Hunterdon': '-75.195496,40.339427,-74.702703,40.787729', 'Mercer': '-74.942909,40.136771,-74.481095,40.424156', 'Middlesex': '-74.631036,40.251122,-74.203605,40.608683', 'Monmouth': '-74.614576,40.07913,-73.88506,40.519095', 'Morris': '-74.889819,40.648447,-74.266591,41.086114', 'Ocean': '-74.553105,39.475198,-73.965414,40.172402', 'Passaic': '-74.503212,40.819938,-74.106786,41.203241', 'Salem': '-75.563586,39.355089,-75.061854,39.796368', 'Somerset': '-74.797881,40.373733,-74.401742,40.7

In [ ]:
for county in NJ_counties_dict.keys():
  print (county)

Atlantic
Bergen
Burlington
Camden
Cape May
Cumberland
Essex
Gloucester
Hudson
Hunterdon
Mercer
Middlesex
Monmouth
Morris
Ocean
Passaic
Salem
Somerset
Sussex
Union
Warren


In [ ]:
df3 = pd.read_csv("NJ_Muni_Bbox_Extents.csv")

In [ ]:
df3.head()
NJ_muni = dict(zip(df3.MUN_LABEL, df3.bbox))
print(NJ_muni)

{'Deerfield Township': '-75.1736,39.4281,-75.0813,39.5187', 'Hopewell Township': '-74.9434,40.2773,-74.7129,40.4242', 'Buena Borough': '-74.9855,39.5056,-74.9140,39.5563', 'Port Republic City': '-74.5159,39.4983,-74.4417,39.5591', 'Lower Alloways Creek Township': '-75.5452,39.3792,-75.3503,39.5565', 'Upper Deerfield Township': '-75.2703,39.4310,-75.1628,39.5594', 'Vineland City': '-75.0876,39.3771,-74.8963,39.5694', 'Elsinboro Township': '-75.5340,39.4941,-75.4593,39.5783', 'Newfield Borough': '-75.0344,39.5373,-75.0023,39.5606', 'Quinton Township': '-75.4604,39.4707,-75.3068,39.5796', 'Beach Haven Borough': '-74.2672,39.5470,-74.2295,39.5853', 'Salem City': '-75.4942,39.5547,-75.4529,39.5839', 'Hamilton Township': '-74.7531,40.1365,-74.5878,40.2710', 'Buena Vista Township': '-74.9727,39.4203,-74.8252,39.6023', 'Cape May Point Borough': '-74.9717,38.9310,-74.9598,38.9431', 'West Cape May Borough': '-74.9520,38.9325,-74.9211,38.9508', 'Cape May City': '-74.9387,38.9284,-74.8635,38.9522'

In [ ]:
df3.head()
NJ_alt_muni = dict(zip(df3.NAME, df3.bbox))
print(NJ_alt_muni)

{'Deerfield Township': '-75.1736,39.4281,-75.0813,39.5187', 'Hopewell Township': '-74.9434,40.2773,-74.7129,40.4242', 'Buena Borough': '-74.9855,39.5056,-74.9140,39.5563', 'Port Republic': '-74.5159,39.4983,-74.4417,39.5591', 'Lower Alloways Creek Township': '-75.5452,39.3792,-75.3503,39.5565', 'Upper Deerfield Township': '-75.2703,39.4310,-75.1628,39.5594', 'Vineland': '-75.0876,39.3771,-74.8963,39.5694', 'Elsinboro Township': '-75.5340,39.4941,-75.4593,39.5783', 'Newfield Borough': '-75.0344,39.5373,-75.0023,39.5606', 'Quinton Township': '-75.4604,39.4707,-75.3068,39.5796', 'Beach Haven Borough': '-74.2672,39.5470,-74.2295,39.5853', 'Salem': '-75.4942,39.5547,-75.4529,39.5839', 'Hamilton Township': '-74.7531,40.1365,-74.5878,40.2710', 'Buena Vista Township': '-74.9727,39.4203,-74.8252,39.6023', 'Cape May Point Borough': '-74.9717,38.9310,-74.9598,38.9431', 'West Cape May Borough': '-74.9520,38.9325,-74.9211,38.9508', 'Cape May': '-74.9387,38.9284,-74.8635,38.9522', 'Wildwood Crest Bo

**Matching municipalities**

In [ ]:
muni_count = {}
#index = 0
for key, value in place_names.items():
  #index += 1
  match = []
  multi_term = value.split("|")
  for term in multi_term:  
    for muni in NJ_muni.keys():
      if muni in term:
        match.append(muni)
  unique = list(set(match))
  if len(unique) == 1:
    for x in match:
      muni_count[key] = [value, len(unique), x, NJ_muni[x]]
  else:
    muni_count[key] = [value, len(unique), "", ""]
print (len(muni_count))

4579


In [ ]:
muni_df = pd.DataFrame.from_dict(muni_count, orient = "index")
muni_df.to_csv('MatchingMunicipalities.csv')
files.download("MatchingMunicipalities.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

**Matching Counties**

In [ ]:
count = {}
#index = 0
for key, value in place_names.items():
  #index += 1
  match = []
  multi_term = value.split("|")
  for term in multi_term:
    for county in NJ_counties_dict.keys():
      if (county + " County") in term:
      #if county in term:
        match.append(county)
  unique = list(set(match))
  if len(unique) == 1:
    for x in match:
      count[key] = [value, len(unique), x, NJ_counties_dict[x]]
  else:
    count[key] = [value, len(unique), "", ""]
print (len(count))

4579


In [ ]:
new_df = pd.DataFrame.from_dict(count, orient = "index")
new_df.to_csv('MatchingCounties.csv')
files.download("MatchingCounties.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

**Combining matching municipalities and counties**

In [ ]:
count = {}
for key, value in place_names.items():
  muni_match = []
  multi_term = value.split("|")
  for term in multi_term:  
    for muni in NJ_muni.keys():
      if muni in term:
        muni_match.append(muni)
  muni_unique = list(set(muni_match))
  if len(muni_unique) == 1:
    for x in muni_match:
      count[key] = [value, len(muni_unique), x, NJ_muni[x],"","",""]
  else:
    county_match = []
    for term in multi_term:
      for county in NJ_counties_dict.keys():
        if (county + " County") in term:
          county_match.append(county)
    county_unique = list(set(county_match))
    if len(county_unique) == 1:
      for y in county_match:
        count[key] = [value, len(muni_unique), "", "", len(county_unique), y, NJ_counties_dict[y]]
    else:
      count[key] = [value, len(muni_unique), "", "", len(county_unique), "", ""]
print (len(count))


4579


**Additional attempts to match municipalities** 

In [ ]:
count = {}
alt_count ={}
unmatched = {}
for key, value in place_names.items():
  muni_match = []
  multi_term = value.split("|")
  for term in multi_term:  
    for muni in NJ_muni.keys():
      if muni in term:
        muni_match.append(muni)
  muni_unique = list(set(muni_match))
  if len(muni_unique) == 1:
    for x in muni_match:
      count[key] = [value, "Municipal", x, NJ_muni[x]]
  else:
    county_match = []
    for term in multi_term:
      for county in NJ_counties_dict.keys():
        if (county + " County") in term:
          county_match.append(county)
    county_unique = list(set(county_match))
    if len(county_unique) == 1:
      for y in county_match:
        count[key] = [value, "County", y, NJ_counties_dict[y]]
    else:
      alt_muni_match = []
      for term in multi_term:  
        for alt_muni in NJ_alt_muni.keys():
          if (alt_muni + ", New Jersey") in term:
            alt_muni_match.append(alt_muni)
          else:
            alt_muni_not = alt_muni.replace(" Township", "")
            alt_muni_nob = alt_muni_not.replace(" Borough", "")
            if (alt_muni_nob + ", New Jersey") in term:
              alt_muni_match.append(alt_muni)
      alt_muni_unique = list(set(alt_muni_match))
      if len(alt_muni_unique) == 1:
        if len(county_unique) < 2:
          for z in alt_muni_match:
            alt_count[key] = [value, "Municipal", z, NJ_alt_muni[z]]
      else:
        unmatched[key] = [value, len(muni_unique), len(county_unique)]
print (len(count))
print (len(alt_count))
print (len(unmatched))

3319
303
942


**Code to download resulting CSVs**

In [ ]:
combine_df = pd.DataFrame.from_dict(count, orient = "index")
combine_df.to_csv('MatchingMuniAndCounties.csv')
files.download("MatchingMuniAndCounties.csv")

In [ ]:
print(len(unmatched))
unmatched_df = pd.DataFrame.from_dict(unmatched, orient = "index")
unmatched_df.to_csv('UnMatchedPlaces.csv')
files.download("UnMatchedPlaces.csv")

In [ ]:
alt_matched_df = pd.DataFrame.from_dict(alt_count, orient = "index")
alt_matched_df.to_csv('Alternative_MatchedPlaces.csv')
files.download("Alternative_MatchedPlaces.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Work still to do:
*   Look at Kalan's process for matching with Geonames
*   Look for more opportunities with unmatched places and better matching for records that only matched with county


Total records: 4579

*   1257 don't have "[Name] County"  (27%)
*   3168 have only one "[Name] County" (69%)
*   154 has more than one "[Name] County" (3%)

---

*   515 don't have one of the named counties in their keyword list  (11%)
*   3658 have only one of the named counties in their keyword list  (80%)
*   406 have more than one of the named counties in their keyword list (9%)



---
combined county/muni 9/29/2022 end of day

1137 no match muni or county (24%)
789 have only one match MUNI_LABEL (17%)
2530 have only one "[Name] County" (55%)
123 match with more than one "[Name] County" (2-3%)